app/models/ems_refresh/save_inventory_container.rb

```diff
@@ -16,20 +31,83 @@ def save_ems_container_inventory(ems, hashes, target = nil)
     ems.save!
   end
 
-  def save_container_projects_inventory(ems, hashes, target = nil)
-    return if hashes.nil?
-    target = ems if target.nil?
-
-    ems.container_projects.reset
-    deletes = if target.kind_of?(ExtManagementSystem)
-                :use_association
-              else
-                []
-              end
+  def initialize_inventory_collections(ems)
+    @inv_collections = {}
+    @inv_collections[:container_projects] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerProject,
+      :parent => ems,
+      :builder_params => {:ems_id => ems.id},
+      :association => :container_projects
+    )
+    @inv_collections[:container_quotas] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerQuota,
+      :parent => ems,
+      :builder_params => {:ems_id => ems.id},
+      :association => :container_quotas,
+      #:arel => ContainerQuota.joins(:container_project).where(:container_projects => {:ems_id => ems.id}),
+    )
+    @inv_collections[:container_quota_items] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerQuotaItem,
+      :parent => ems,
+      :association => :container_quota_items,
+      #:arel => ContainerQuotaItem.joins(:container_quota => :container_project).where(:container_projects => {:ems_id => ems.id}),
+      :manager_ref => [:container_quota, :resource],
+    )
+    @inv_collections[:container_nodes] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerNode,
+      :parent => ems,
+      :builder_params => {:ems_id => ems.id},
+      :association => :container_nodes,
+    )
+    # ContainerCondition is polymorphic child of ContainerNode & ContainerGroup
+    @inv_collections[:container_node_conditions] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerCondition,
+      :parent => ems,
+      :association => :container_node_conditions,
+      #:arel => ContainerCondition.joins(
+      #  'INNER JOIN "container_nodes" ON "container_conditions"."container_entity_id" = "container_nodes"."id"'
+      #).where(:container_entity_type => 'ContainerNode',
+      #        :container_nodes => {:ems_id => ems.id}),
+      :manager_ref => [:container_entity, :name],
+    )
+
+    # polymorphic child of ContainerNode & ContainerImage
+    # TODO these are dumb, find way to share same InventoryCollections via both.
```

Comment:
Right, so to be effective, the InventoryCollection is basically a DB table. You should be able to get what is in the table by one query (scoped on EMS) and the table should have a DB unique index (I'll be adding those shortly, since those are missing almost everywhere)

So for the particular issue you have a 1 query to get them all can be done via union, to get ComputerSystem of both Hardware and OperatingSystem, for Ems. 

Or as we were talking with @agrare, to have simpler queries, we can add [:ems_id, :ems_ref] to all these problematic tables (which would become actual DB unique index in most tables), while you would build the ems_ref to be unique under ems. Then the queries should be easy, since you can use the :ems_id to easily load all under Ems and :ems_ref to filter our few in a case of the targeted refresh.  

So I would go with [:ems_id, :ems_ref] for ComputerSystem and Labels (not sure about tags, there is a special saving logic we need to refactor, most likely those should get out of refresh and be post refresh). It's uglier from the modeling point of view, but I assume nobody is keen to write union queries. :-) @agrare agree?

note: now without the unique DB index, since for that, we will need a plenty of data migrations